### PR TITLE
Use SlicesIterator for ArrayData Equality

### LINKED ISSUE #1600: ArrayData Equality Slice Iterator
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

A lot of the ArrayData equality implementation iterate through row-wise in the event of nulls, it would be advantageous to instead iterate over valid ranges, for example, using `SlicesIterator`. I originally planned to do this as part of #1589 but that PR already is rather large so creating as ticket and will follow up later

**Describe the solution you'd like**

Should use `SlicesIterator` to amortize the cost of performing equality comparisons

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1600.

# Rationale for this change
 

This improves the performance of long array with less nulls mostly, e.g. the added benchmark:

```
equal_51200             time:   [72.533 µs 72.653 µs 72.810 µs]     
                        change: [-30.087% -30.012% -29.938%] (p = 0.00 < 0.05)
                        Performance has improved.   
```

For array with many nulls, this actually causes regression when using `SlicesIterator`, so I only use `SlicesIterator` for the cases with less nulls.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by tustvold — COMMENTED** [review]
Perhaps this could be simplified by using try_for_each_valid_idx?

**Comment by viirya** [comment]
> Perhaps this could be simplified by using try_for_each_valid_idx?

Tried with `try_for_each_valid_idx`, but the benchmarks look not good:

```
equal_nulls_512         time:   [1.8361 µs 1.8471 µs 1.8602 µs]
                        change: [+171.23% +172.28% +173.48%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 23 outliers among 100 measurements (23.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  3 (3.00%) high mild
  18 (18.00%) high severe

equal_51200             time:   [303.39 µs 303.51 µs 303.67 µs]
                        change: [+191.52% +191.71% +191.91%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
```

`try_for_each_valid_idx` comes to compare each element so it doesn't help on the long array with few nulls (i.e., `equal_51200`), it also brings quite regression on many null case.

File: arrow-data/src/slices_iterator.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+/// An iterator of `(usize, usize)` each representing an interval
+/// `[start, end)` whose slots of a bitmap [Buffer] are true. Each
+/// interval corresponds to a contiguous region of memory to be
+/// "taken" from an array to be filtered.
+///
+/// ## Notes:
+///
+/// Only performant for filters that copy across long contiguous runs
+#[derive(Debug)]
+pub struct SlicesIterator<'a>(BitSliceIterator<'a>);
```
Comment (tustvold) [inline arrow-data/src/slices_iterator.rs:26]: Moving this is a breaking change, as it is exposed publicly - https://docs.rs/arrow-select/latest/arrow_select/filter/struct.SlicesIterator.html

File: arrow-data/src/slices_iterator.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+impl<'a> SlicesIterator<'a> {
+    pub fn new_from_buffer(values: &'a Buffer, offset: usize, len: usize) -> Self {
+        Self(BitSliceIterator::new(values, offset, len))
```
Comment (tustvold) [inline arrow-data/src/slices_iterator.rs:30]: At this point why not just use `BitSliceIterator`?

File: arrow-data/src/equal/primitive.rs [hunk 3]
```diff
@@ -15,13 +15,17 @@
...
 use std::mem::size_of;
 
-use crate::data::{contains_nulls, ArrayData};
-use arrow_buffer::bit_util::get_bit;
+use crate::data::ArrayData;
+use crate::slices_iterator::SlicesIterator;
 
 use super::utils::equal_len;
 
+pub(crate) const NULL_SLICES_SELECTIVITY_THRESHOLD: f64 = 0.4;
```
Comment (tustvold) [inline arrow-data/src/equal/primitive.rs:27]: How did you come up with 0.4, not saying it is a bad choice, just curious

**Review by tustvold — COMMENTED** [review]
Could we possibly back out the changes to `SlicesIterator` and just use `BitSliceIterator`, then I think this is good to go :+1:

File: arrow-data/src/equal/primitive.rs (hunk 3 above)
Comment (viirya) [reply arrow-data/src/equal/primitive.rs:27]: It came from benchmarking. `equal_nulls_512` has 0.5 null density, and this doesn't improve it. So I pick 0.4 as the threshold.

**Review by tustvold — APPROVED** [review]

**Pull request merged**